Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -742,6 +739,80 @@ public void run() {
     return resultRef.get();
   }
 
+  public void search_findElementReferences(@NotNull final String filePath,
+                                           final int offset,
+                                           @NotNull final Processor<SearchResult> processor) {
+    final String searchId;
+    synchronized (myLock) {
+      if (myServer == null) return;
+      final AnalysisServer server = myServer;
+
+      final Ref<String> searchIdRef = new Ref<String>();
+      final Semaphore semaphore = new Semaphore();
+
+      semaphore.down();
+      final boolean ok = runInPooledThreadAndWait(new Runnable() {
+        @Override
+        public void run() {
+          server.search_findElementReferences(filePath, offset, true, new FindElementReferencesConsumer() {
+            @Override
+            public void computedElementReferences(String searchId, Element element) {
+              searchIdRef.set(searchId);
+              semaphore.up();
+            }
+
+            @Override
+            public void onError(RequestError requestError) {
+              semaphore.up();
+            }
+          });
+        }
+      }, "search_findElementReferences(" + filePath + ", " + offset + ")", SEND_REQUEST_TIMEOUT);
+
+      if (!ok) {
+        stopServer();
+        return;
+      }
+
+      final long t0 = System.currentTimeMillis();
+      semaphore.waitFor(FIND_ELEMENT_REFERENCES_TIMEOUT);
+
+      if (semaphore.tryUp()) {
+        LOG.info("search_findElementReferences() took too long for file " +
+                 filePath +
+                 "@" +
+                 offset +
+                 ": " +
+                 (System.currentTimeMillis() - t0) +
+                 "ms");
+        return;
+      }
+
+      searchId = searchIdRef.get();
+    }
+
+    while (true) {
```

Comment:
In some cases this loop becomes infinite. For example take simple Dart file and search for function usages:
`foooo(){}`
In this case server gives zero notifications via `computedSearchResults()`. Here's full trace:
`
debugStream: 1437741190475 => {"id":"22","method":"search.findElementReferences","params":{"file":"C:/flex_projects/untitled5/dart/bin/dart.dart","offset":0,"includePotential":true},"clientRequestTime":1437741190475}`
`
debugStream: 1437741190480 <= {"id":"22","result":{}}
`

Any ideas how to fix it?